src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs

```diff
@@ -693,6 +693,39 @@ public async Task WebApplicationConfiguration_HostFilterOptionsAreReloadable()
             Assert.Contains("NewHost", options.AllowedHosts);
         }
 
+        [Fact]
+        public void CanResolveIConfigurationBeforeBuildingApplication()
+        {
+            var builder = WebApplication.CreateBuilder();
+            var sp = builder.Services.BuildServiceProvider();
+
+            var config = sp.GetService<IConfiguration>();
+            Assert.NotNull(config);
+            Assert.Same(config, builder.Configuration);
+
+            var app = builder.Build();
+
+            // These are different
+            Assert.NotSame(app.Configuration, builder.Configuration);
+        }
+
+        [Fact]
+        public void ManuallyAddingConfigurationAsServiceWorks()
+        {
+            var builder = WebApplication.CreateBuilder();
+            builder.Services.AddSingleton<IConfiguration>(builder.Configuration);
+            var sp = builder.Services.BuildServiceProvider();
+
+            var config = sp.GetService<IConfiguration>();
+            Assert.NotNull(config);
+            Assert.Same(config, builder.Configuration);
+
+            var app = builder.Build();
+
+            // These are different
+            Assert.NotSame(app.Configuration, builder.Configuration);
```

Comment:
This is kinda annoying, no? Can we keep the `builder.Configuration` registration if it's added manually?

Reply:
I can make `WebApplication.Configuration` and `WebApplicationBuilder.Configuration` match, but I can't make them match the final `IConfiguration` instance. The problem is that we need to chain the `ConfigurationManager` to the final `IConfiguration` so we can see the providers added by unit tests. There's no way to get the `IConfiguration` from the built application other than via the `IServiceProvider`.